orderer/consensus/follower/follower_chain.go

```diff
@@ -51,16 +143,128 @@ func (*Chain) Errored() <-chan struct{} {
 }
 
 func (c *Chain) Start() {
-	//TODO skeleton
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.started || c.stopped {
+		c.logger.Debugf("Not starting because: started=%v, stopped=%v", c.started, c.stopped)
+		return
+	}
+
+	c.started = true
+
+	go c.run()
+
+	c.logger.Info("Started")
 }
 
 func (c *Chain) Halt() {
-	//TODO skeleton
+	c.halt()
+
+	select {
+	case <-c.doneChan:
+	}
+}
+
+func (c *Chain) halt() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.stopped {
+		c.logger.Debug("Already stopped")
+		return
+	}
+	c.stopped = true
+	close(c.stopChan)
+	c.logger.Info("Stopped")
 }
 
-// StatusReport returns the ClusterRelation & Status
+// StatusReport returns the ClusterRelation & Status.
 func (c *Chain) StatusReport() (types.ClusterRelation, types.Status) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	clusterRelation := types.ClusterRelationFollower
+	if c.joinBlock != nil {
+		if err := c.amIInChannel(c.joinBlock); err == nil {
+			clusterRelation = types.ClusterRelationMember
+		}
+	}
+
 	status := types.StatusActive
-	//TODO if (height is >= join-block.height) return "active"; else return "onboarding"
-	return types.ClusterRelationFollower, status
+	if (c.joinBlock != nil) && (c.joinBlock.Header.Number >= c.support.Height()) {
+		status = types.StatusOnBoarding
+	}
+
+	return clusterRelation, status
+}
+
+func (c *Chain) run() {
+	c.logger.Debug("The follower.Chain puller goroutine is starting")
+	pullRetryTicker := time.NewTicker(500 * time.Millisecond) // TODO move constant to options
+
+	err := c.pull()
+
+runLoop:
+	for err != nil {
+		c.logger.Debugf("Pull failed, going to try again in %v; error: %s", pullRetryTicker, err)
+
+		select {
+		case <-c.stopChan:
+			c.logger.Debug("Received a stop signal")
+			break runLoop
+
+		case <-pullRetryTicker.C:
+			err = c.pull()
+		}
+	}
+
+	pullRetryTicker.Stop()
+	close(c.doneChan)
+	c.logger.Debug("The follower.Chain puller goroutine is exiting")
+}
+
+// pull blocks from other orderers, until the join block, or until a config block the indicates the orderer has become
+// a member pf the cluster. When the follower.Chain's job is done, this method halts, triggers the creation of a new
+// consensus.Chain, and returns nil.
+func (c *Chain) pull() error {
+	if c.joinBlock != nil {
+		err := c.pullUpToJoin()
+		if err != nil {
+			return errors.Wrap(err, "failed to pull up to join block")
+		}
+
+		// TODO remove the join-block from the file repo
+	} else {
+		err := c.pullAfterJoin()
+		if err != nil {
+			return errors.Wrap(err, "failed to pull after join block")
+		}
+	}
+
+	// Trigger creation of a new chain. This will regenerate 'support' from the tip of the ledger, and will
+	// start either a follower.Chain or an etcdraft.Chain, depending on whether the orderer is a follower or a
+	// member of the cluster.
+	c.halt()
+	if c.chainCreationCallback != nil {
```

Comment:
consider checking at initialization and just assining a no-op if nil and anyway invoke it here anyway

Reply:
That would just add lines to the code while at the same time making it less readable... I mean the `if c.chainCreationCallback != nil {...}` is still there, but the noop now has to be defined.... what is the benefit?